armory/metrics/perturbation.py

```diff
@@ -0,0 +1,340 @@
+        super().__setattr__(name, function)
+
+    def __delattr__(self, name):
+        raise ValueError("Deletion not allowed")
+
+    def _names(self):
+        return sorted(x for x in self.__dict__ if not x.startswith("_"))
+
+    def __contains__(self, name):
+        return name in self._names()
+
+    def __repr__(self):
+        """
+        Show the existing non-underscore names
+        """
+        return str(self._names())
+
+    def __iter__(self):
+        for name in self._names():
+            yield name, self[name]
+
+    def __getitem__(self, name):
+        if not hasattr(self, name):
+            raise KeyError(name)
+        return getattr(self, name)
+
+    def __setitem__(self, name, function):
+        setattr(self, name, function)
+
+
+element = MetricNameSpace()
+batch = MetricNameSpace()
+
+
+def as_batch(element_metric):
+    """
+    Return a batchwise metric function from an elementwise metric function
+    """
+
+    @functools.wraps(element_metric)
+    def wrapper(x_batch, x_adv_batch, **kwargs):
+        x_batch = list(x_batch)
+        x_adv_batch = list(x_adv_batch)
+        if len(x_batch) != len(x_adv_batch):
+            raise ValueError(
+                f"len(a_batch) {len(x_batch)} != len(b_batch) {len(x_adv_batch)}"
+            )
+        y = []
+        for x, x_adv in zip(x_batch, x_adv_batch):
+            y.append(element_metric(x, x_adv, **kwargs))
+        try:
+            y = np.array(y)
+        except ValueError:
+            # Handle ragged arrays
+            y = np.array(y, dtype=object)
+        return y
+
+    if wrapper.__doc__ is None:
+        log.warning(f"{element_metric.__name__} has no doc string")
+        wrapper.__doc__ = ""
+    wrapper.__doc__ = "Batch version of:\n" + wrapper.__doc__
+    wrapper.__name__ = "batch_" + wrapper.__name__
+    # note: repr(wrapper) defaults to the element_metric, not __name__
+    # See: https://stackoverflow.com/questions/10875442/possible-to-change-a-functions-repr-in-python
+    return wrapper
+
+
+def batchwise(batch_metric, name=None):
+    """
+    Register a batch metric and register a batchwise version of it
+    """
+    if name is None:
+        name = batch_metric.__name__
+    setattr(batch, name, batch_metric)
+
+
+def elementwise(element_metric, name=None):
+    """
+    Register a element metric and register a batchwise version of it
+    """
+    if name is None:
+        name = element_metric.__name__
+    setattr(element, name, element_metric)
+    batch_metric = as_batch(element_metric)
+    batchwise(batch_metric, name=name)
+    return element_metric
+
+
+def numpy(function):
+    """
+    Ensures args (but not kwargs) are passed in as numpy vectors
+    It casts them to to a common data type (complex, float, int) if possible
+    And it ensures that they have the same shape
+    """
+
+    @functools.wraps(function)
+    def wrapper(x, x_adv, **kwargs):
+        # TODO: cast from Torch tensor and TF tensor to numpy?
```

Comment:
is the TODO here bc we were worried that `np.asarray` might not work for torch and tf? If it does, should we remove the comment (otherwise, should we address it)?

Reply:
I just checked, and it looks like `np.asarray` works fine for torch and tf tensors. So I'm going to remove the TODO.

Reply:
Done.